openhands/server/session/session.py

```diff
@@ -133,7 +134,19 @@ async def initialize_agent(
             self.logger.info(f'Enabling default condenser: {default_condenser_config}')
             agent_config.condenser = default_condenser_config
 
-        agent = Agent.get_cls(agent_cls)(llm, agent_config)
+        mcp_clients = await create_mcp_clients(
+            self.config.mcp.sse.mcp_servers,
+            self.config.mcp.stdio.commands,
+            self.config.mcp.stdio.args,
+        )
+        mcp_tools = convert_mcp_clients_to_tools(mcp_clients)
+        if mcp_tools:
+            agent = Agent.get_cls(agent_cls)(llm, agent_config, mcp_tools)
+        else:
+            agent = Agent.get_cls(agent_cls)(llm, agent_config)
+        # close all mcp clients after extracting tools
+        for mcp_client in mcp_clients:
+            await mcp_client.disconnect()
```

Comment:
ditto -- we might not need to explicitly create clients here to get a list of tools? If it is absolutely necessary to start client here, can we just make a helper function that takes in config from MCP, and then return a list of tools. So we can reuse this in multiple places?

Reply:
Yea this is kinda like a "trick" to let the agent know about the tools. Afaik, we can only get the list of tools after successfully connecting to an MCP server.

My main goal here is to dynamically get the tools from the list of servers without knowing other contexts.

Reply:
Sounds good! Maybe we could later refactor this in the a function, so we can use them in create_agent too!